### ISSUE TITLE
Make JuliaOJIT non-eager

### ISSUE THREAD

**Issue by vchuravy** [JuliaLang/julia#52130]
On slack a new TTFX example emerged: `julia --startup-file=no -e 'using LinearAlgebra; sqrt([1 2; 3 4])'` which takes >10s.

Quickly profiling it shows that this is entirely compilation bound, with
~45% in LLVM optimizations, 20% in inference, 3.5% in Codegen

![image](https://github.com/JuliaLang/julia/assets/145258/5ea57441-52a5-4f4e-b1e4-e2cbda43bea2)

One issue is that we are not exploiting the parallelism present since `JuliaOJIT::addModule` is eager.

https://github.com/JuliaLang/julia/blob/c15f4e9ed5540b8a64558c985a4ffda50151353c/src/jitlayers.cpp#L1866-L1867

Once we are on LLVM 16 we should take another look if we can use JITLink everywhere.

**Comment by vchuravy** [comment]
@pchintalapudi is my recollection correct? With JITLink we would be using a memory manager that would handle this better?

**Comment by gbaraldi** [comment]
I'm not sure making it non eager would make that big of a difference. The lazy part of it is less for parallelization and more for not compiling code that doesn't get run, but here I think it will get run right away.

**Comment by pchintalapudi** [comment]
I agree with Gabriel; I doubt this particular example would be better suited for parallel compilation. I could see tiered/queued compilation being more helpful for this case. 

That being said, this looks so simple that I feel tweaking the pass pipeline might fix the issue, if there's something that's causing LLVM to spend 5 seconds on this one function.
